From 4f04d8b6c9f9a7c34ce5c9727e608aea43c9b310 Mon Sep 17 00:00:00 2001
From: Wang Han <wanghan1995315@gmail.com>
Date: Sun, 13 May 2018 17:56:39 +0200
Subject: [PATCH 09/27] Keyguard: Fix ConcurrentModificationException in
 KeyguardUpdateMonitor

Seems there is some races here, so use ConcurrentHashMap to avoid
any issues.

Crash logs:
E AndroidRuntime: FATAL EXCEPTION: main
E AndroidRuntime: Process: com.android.systemui, PID: 9210
E AndroidRuntime: java.util.ConcurrentModificationException
E AndroidRuntime: 	at java.util.HashMap$HashIterator.nextNode(HashMap.java:1441)
E AndroidRuntime: 	at java.util.HashMap$EntryIterator.next(HashMap.java:1475)
E AndroidRuntime: 	at java.util.HashMap$EntryIterator.next(HashMap.java:1475)
E AndroidRuntime: 	at com.android.keyguard.KeyguardUpdateMonitor.sendUpdates(KeyguardUpdateMonitor.java:1685)
E AndroidRuntime: 	at com.android.keyguard.KeyguardUpdateMonitor.registerCallback(KeyguardUpdateMonitor.java:1665)
E AndroidRuntime: 	at com.android.keyguard.KeyguardSimPinView.onAttachedToWindow(KeyguardSimPinView.java:175)
E AndroidRuntime: 	at android.view.View.dispatchAttachedToWindow(View.java:17445)
E AndroidRuntime: 	at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3326)
E AndroidRuntime: 	at android.view.ViewGroup.addViewInner(ViewGroup.java:4977)
E AndroidRuntime: 	at android.view.ViewGroup.addView(ViewGroup.java:4768)
E AndroidRuntime: 	at android.widget.ViewAnimator.addView(ViewAnimator.java:183)
E AndroidRuntime: 	at android.view.ViewGroup.addView(ViewGroup.java:4708)
E AndroidRuntime: 	at android.view.ViewGroup.addView(ViewGroup.java:4681)
E AndroidRuntime: 	at com.android.keyguard.KeyguardSecurityContainer.getSecurityView(KeyguardSecurityContainer.java:147)
E AndroidRuntime: 	at com.android.keyguard.KeyguardSecurityContainer.showSecurityScreen(KeyguardSecurityContainer.java:379)
E AndroidRuntime: 	at com.android.keyguard.KeyguardSecurityContainer.showPrimarySecurityScreen(KeyguardSecurityContainer.java:310)
E AndroidRuntime: 	at com.android.keyguard.KeyguardHostView.showPrimarySecurityScreen(KeyguardHostView.java:159)
E AndroidRuntime: 	at com.android.systemui.statusbar.phone.KeyguardBouncer.show(KeyguardBouncer.java:108)
E AndroidRuntime: 	at com.android.systemui.statusbar.phone.StatusBarKeyguardViewManager.showBouncerOrKeyguard(StatusBarKeyguardViewManager.java:166)
E AndroidRuntime: 	at com.android.systemui.statusbar.phone.StatusBarKeyguardViewManager.reset(StatusBarKeyguardViewManager.java:253)
E AndroidRuntime: 	at com.android.systemui.statusbar.phone.StatusBarKeyguardViewManager.reset(StatusBarKeyguardViewManager.java:264)
E AndroidRuntime: 	at com.android.systemui.keyguard.KeyguardViewMediator.handleReset(KeyguardViewMediator.java:1923)
E AndroidRuntime: 	at com.android.systemui.keyguard.KeyguardViewMediator.-wrap13(Unknown Source:0)
E AndroidRuntime: 	at com.android.systemui.keyguard.KeyguardViewMediator$4.handleMessage(KeyguardViewMediator.java:1533)
E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:164)
E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:6494)
E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:440)
E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)

Change-Id: Icb77e7ac172b0d186082cadef9bdd7e2e6165230
---
 .../src/com/android/keyguard/KeyguardUpdateMonitor.java        | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/packages/SystemUI/src/com/android/keyguard/KeyguardUpdateMonitor.java b/packages/SystemUI/src/com/android/keyguard/KeyguardUpdateMonitor.java
index 445b619d93d..e5eb78f3e1f 100644
--- a/packages/SystemUI/src/com/android/keyguard/KeyguardUpdateMonitor.java
+++ b/packages/SystemUI/src/com/android/keyguard/KeyguardUpdateMonitor.java
@@ -88,6 +88,7 @@ import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
 
 /**
  * Watches for updates that may be interesting to the keyguard, and provides
@@ -181,7 +182,7 @@ public class KeyguardUpdateMonitor implements TrustManager.TrustListener {
     private static KeyguardUpdateMonitor sInstance;
 
     private final Context mContext;
-    HashMap<Integer, SimData> mSimDatas = new HashMap<Integer, SimData>();
+    ConcurrentHashMap<Integer, SimData> mSimDatas = new ConcurrentHashMap<Integer, SimData>();
     HashMap<Integer, ServiceState> mServiceStates = new HashMap<Integer, ServiceState>();
 
     private int mRingMode;
-- 
2.17.1

